src/persist/src/indexed/background.rs

```diff
@@ -140,49 +140,105 @@ impl<B: Blob> Maintainer<B> {
+            updates.append(&mut first_updates);
+            updates.append(&mut second_updates);
+            for ((_, _), ts, _) in updates.iter_mut() {
+                ts.advance_by(desc.since().borrow());
+            }
+            differential_dataflow::consolidation::consolidate_updates(&mut updates);
+            for ((key, val), time, diff) in updates.iter() {
+                builder.push(((key, val), *time, *diff));
+                if builder.num_filled() > 0 {
+                    let filled = builder.take_filled();
+                    for batch in filled {
+                        let new_batch = BlobTraceBatch {
+                            desc: desc.clone(),
+                            updates: vec![batch],
+                        };
+                        let new_key = Arrangement::new_blob_key();
+                        let (format, size_bytes) =
+                            blob.set_trace_batch(new_key.clone(), new_batch)?;
+                        keys.push(new_key);
+                        formats.push(format);
+                        new_size_bytes += size_bytes;
+                    }
+                }
+            }
+
+            updates.clear();
         }
 
-        differential_dataflow::consolidation::consolidate_updates(&mut updates);
-
-        let updates = updates.iter().collect::<ColumnarRecordsVec>().into_inner();
-        let new_batch = BlobTraceBatch {
-            desc: desc.clone(),
-            updates,
-        };
-
-        let merged_key = Arrangement::new_blob_key();
-        let (format, size_bytes) = blob.set_trace_batch(merged_key.clone(), new_batch)?;
+        // Grab the last columnar record if one remains
+        let finished = builder.finish();
+        for batch in finished {
+            let new_batch = BlobTraceBatch {
+                desc: desc.clone(),
+                updates: vec![batch],
+            };
+            let new_key = Arrangement::new_blob_key();
+            let (format, size_bytes) = blob.set_trace_batch(new_key.clone(), new_batch)?;
+            keys.push(new_key);
+            formats.push(format);
+            new_size_bytes += size_bytes;
+        }
 
         // Only upgrade the compaction level if we know this new batch represents
         // an increase in data over both of its parents so that we know we need
         // even more additional batches to amortize the cost of compacting it in
         // the future.
-        let merged_level = if size_bytes > first.size_bytes && size_bytes > second.size_bytes {
-            first.level + 1
-        } else {
-            first.level
-        };
+        let merged_level =
+            if new_size_bytes > first.size_bytes && new_size_bytes > second.size_bytes {
+                first.level + 1
+            } else {
+                first.level
+            };
 
         let merged = TraceBatchMeta {
-            key: merged_key,
-            format,
+            key: keys,
+            format: formats[0],
```

Comment:
when we do this for real, I think we'll want to pass the format to set_trace_batch instead so we don't have to deal with this